addons/binding/org.openhab.binding.nibeheatpump/src/main/java/org/openhab/binding/nibeheatpump/internal/connection/SerialConnector.java

```diff
@@ -0,0 +1,371 @@
+                }
+
+                // all bytes should be handled, clear buffer for next round
+                context.buffer().clear();
+            }
+
+            logger.debug("Data listener stopped");
+        }
+
+        private byte[] getAllAvailableBytes(InputStream in) throws IOException {
+            ByteArrayOutputStream os = new ByteArrayOutputStream();
+
+            int b;
+            // wait first byte (blocking)
+            if ((b = in.read()) > -1) {
+                byte d[] = new byte[] { (byte) b };
+                os.write(d);
+
+                // read rest of the available bytes
+                final int bufferLen = 100;
+                byte[] buffer = new byte[bufferLen];
+                int available = in.available();
+                if (available > 0) {
+                    int len = in.read(buffer, 0, bufferLen);
+                    if (len > -1) {
+                        os.write(buffer, 0, len);
+                    }
+                }
+
+                os.flush();
+                return os.toByteArray();
+            }
+
+            return null;
+        }
+
+        @Override
+        public void serialEvent(SerialPortEvent event) {
+            try {
+                /*
+                 * See more details from
+                 * https://github.com/NeuronRobotics/nrjavaserial/issues/22
+                 */
+                logger.trace("RXTX library CPU load workaround, sleep forever");
+                sleep(Long.MAX_VALUE);
+            } catch (InterruptedException e) {
+            }
+        }
+    }
+
+    @SuppressWarnings("unused")
+    private void sendNakToNibe() throws IOException {
+        logger.trace("Send data (len=1): 15");
+        out.write(0x15);
+        out.flush();
+    }
+
+    private void sendAckToNibe(byte address) throws IOException {
+        boolean sendack = false;
+
+        if (address == NibeHeatPumpProtocol.ADR_MODBUS40 && conf.sendAckToMODBUS40) {
+            logger.debug("Send ack to MODBUS40 message");
+            sendack = true;
+        } else if (address == NibeHeatPumpProtocol.ADR_SMS40 && conf.sendAckToSMS40) {
+            logger.debug("Send ack to SMS40 message");
+            sendack = true;
+        } else if (address == NibeHeatPumpProtocol.ADR_RMU40 && conf.sendAckToRMU40) {
+            logger.debug("Send ack to RMU40 message");
+            sendack = true;
+        }
+
+        if (sendack) {
+            sendAckToNibe();
+        }
+    }
+
+    private void sendAckToNibe() throws IOException {
+        logger.trace("Send data (len=1): 06");
+        out.write(0x06);
+        out.flush();
+    }
+
+    private void sendDataToNibe(byte[] data) throws IOException {
+        if (logger.isTraceEnabled()) {
+            try {
+                NibeHeatPumpMessage msg = MessageFactory.getMessage(data);
+                logger.trace("Sending msg: {}", msg.toString());
+            } catch (NibeHeatPumpException e) {
+                // do nothing
```

Comment:
It's purely for logging purposes (trace level). I don't want to change binding behaviour when trace level is enabled